design/capm3-remediation-controller-proposal.md

```diff
@@ -0,0 +1,382 @@
+    // +optional
+    LastRebooted *metav1.Time `json:"lastRebooted,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=metal3remediation,scope=Namespaced,categories=cluster-api,shortName=m3r;m3remediation
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+
+// Metal3Remediation is the Schema for the metal3remediations API
+type Metal3Remediation struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec   Metal3RemediationSpec   `json:"spec,omitempty"`
+    Status Metal3RemediationStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// Metal3RemediationList contains a list of Metal3Remediation
+type Metal3RemediationList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata,omitempty"`
+    Items           []Metal3Remediation `json:"items"`
+}
+
+func init() {
+    SchemeBuilder.Register(&Metal3Remediation{}, &Metal3RemediationList{})
+}
+```
+
+New Metal3RemediationTemplate CRD will be added as a part of v1alpha4 API.
+
+```go
+package v1alpha4
+
+import (
+    metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Metal3RemediationTemplateSpec defines the desired state of Metal3RemediationTemplate
+type Metal3RemediationTemplateSpec struct {
+    Template Metal3RemediationTemplateResource `json:"template"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=metal3remediationtemplate,scope=Namespaced,categories=cluster-api
+
+// Metal3RemediationTemplate is the Schema for the metal3remediationtemplate API
+type Metal3RemediationTemplate struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec   Metal3RemediationTemplateSpec   `json:"spec,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// Metal3RemediationTemplateList contains a list of Metal3RemediationTemplate
+type Metal3RemediationTemplateList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata,omitempty"`
+    Items           []Metal3RemediationTemplate `json:"items"`
+}
+
+func init() {
+    SchemeBuilder.Register(&Metal3RemediationTemplate{}, &Metal3RemediationTemplateList{})
+}
+
+// Metal3MachineTemplateResource describes the data needed to create a
+// Metal3Machine from a template
+type Metal3RemediationTemplateResource struct {
+    // Spec is the specification of the desired behavior of the machine.
+    Spec Metal3RemediationSpec `json:"spec"`
+}
+
+```
+
+### Risks and Mitigations
+
+There is a slight possibility that
+[BareMetalHost v1alpha2 API Migration](https://github.com/metal3-io/metal3-docs/pull/101)
+changes the APIs and some rewriting of the Remediation Controller
+code needs to be done after that.
+
+### Work Items
+
+* Implementation of Remediation Controller
```

Comment:
Where will the new controller live?

Reply:
Remediation Controller will be a new controller in CAPM3